docs/proposals/20190709-cluster-spec-crds.md

```diff
@@ -0,0 +1,325 @@
+- The Infrastructure object has its owner set to the Cluster
+
+#### Provisioned
+
+The provider has finished provisioning the infrastructure.
+
+##### Conditions
+- The Cluster object has the `InfrastructureStatusRef` set 
+
+```
+               +----------------+
+               |                |   Cluster object is created
+               |    Pending     |   but has no associated 
+               |                |   Infrastructure
+               +----------------+
+                       |
+                       V
+               +----------------+
+               |                |   Cluster has an associated
+               |  Provisioning  |   infrastructure. Provider is 
+               |                |   provisioning it.
+               +----------------+
+                       |
+                       V
+               +----------------+
+               |                |   Provider has signaled the
+               |  Provisioned   |   end of the provisioning 
+               |                |   
+               +----------------+
+```
+Figure 2. Cluster State transitions
+
+### User Stories [optional]
+
+#### As an infrastructure provider author, I would like to take advantage of the Kubernetes API to provide validation for provider-specific data needed to provision a machine.
+
+#### As an infrastructure provider author, I would like to build a controller to manage provisioning machines using tools of my own choosing.
+
+#### As an infrastructure provider author, I would like to build a controller to manage provisioning clusters without being restricted to a CRUD API.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Role of Cluster Controller
+The Cluster Controller should be the only controller having write permissions to the Cluster objects. Its main responsibility is to 
+- Manage cluster finalizers
+- Set provider's infrastructure object's OwnerRef to Cluster
+- Update the state of the infrastructure provisioning
+
+#### Cluster Controller dynamic watchers
+
+As the provider spec data is no longer inlined in the Cluster object, the Cluster Controller needs to watch for updates to provider specific resources so it can detect events. To achieve this, when the Cluster Controller reconciles a Cluster and detects a reference to a provider-specific infrastructure object, it starts watching this resource using the [dynamic informer](https://godoc.org/k8s.io/client-go/dynamic/dynamicinformer) and a [`handler.EnqueueRequestForOwner`](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/handler#EnqueueRequestForOwner) configured with Cluster as the OwnerType, to ensure it only watches provider objects related to a CAPI cluster.
+
+#### Handling optionality
+
+As both `InfrastructureRef` and `InfrastructureStatusRef` are optional, implementation must be able of differentiating the case when these fields are used by a provider, but have not yet been set, from the case when they are not in use. 
```

Comment:
I'd like to avoid handling optional infrastructure, if the user doesn't want to use the Cluster object it can do so today. If it does I think this field should be required as in Machine.

Reply:
I'm not sure I understand the comment. The optionality si for the InfrastructureRef,  in the same way the existing provider specific spec is optional. It is not related to having the Cluster object or not.